File: clisonix_sdk.ts
/**
 * Clisonix Cloud SDK – TypeScript/JavaScript Client
 * Part of UltraWebThinking / Euroweb
 */

interface ClisonixConfig {
  baseUrl?: string;
  token?: string;
  timeout?: number;
}

interface RequestOptions {
  method?: "GET" | "POST" | "PUT" | "DELETE";
  headers?: Record<string, string>;
  body?: any;
  timeout?: number;
}

export class ClisonixClient {
  private baseUrl: string;
  private token: string | null;
  private refreshToken: string | null;
  private apiKey: string | null;
  private timeout: number;

  constructor(config: ClisonixConfig = {}) {
    this.baseUrl = (config.baseUrl || "https://api.clisonix.com").replace(/\/$/, "");
    this.token = config.token || null;
    this.refreshToken = null;
    this.apiKey = null;
    this.timeout = config.timeout || 30000;
  }

  private getHeaders(extra: Record<string, string> = {}): Record<string, string> {
    const headers: Record<string, string> = {
      "Accept": "application/json",
      "User-Agent": "Clisonix-TypeScript-SDK/1.0",
      ...extra
    };

    if (this.token) {
      headers["Authorization"] = `Bearer ${this.token}`;
    }

    if (this.apiKey) {
      headers["X-API-Key"] = this.apiKey;
    }

    return headers;
  }

  /**
   * POST /auth/login – Login with email and password
   * Automatically stores token, refreshToken, and apiKey
   */
  async login(email: string, password: string): Promise<any> {
    const response: any = await this.request("/auth/login", {
      method: "POST",
      body: { email, password }
    });

    if (response.token) {
      this.token = response.token;
    }
    if (response.refresh_token) {
      this.refreshToken = response.refresh_token;
    }
    if (response.api_key) {
      this.apiKey = response.api_key;
    }

    return response;
  }

  /**
   * POST /auth/refresh – Refresh JWT token using refresh_token
   * Automatically updates stored token
   */
  async refresh(): Promise<any> {
    if (!this.refreshToken) {
      throw new Error("No refresh token available. Call login() first.");
    }

    const response: any = await this.request("/auth/refresh", {
      method: "POST",
      body: { refresh_token: this.refreshToken }
    });

    if (response.token) {
      this.token = response.token;
    }

    return response;
  }

  /**
   * POST /auth/api-key – Create a new API key
   * Requires authentication
   */
  async createApiKey(label: string): Promise<any> {
    if (!this.token) {
      throw new Error("Authentication required. Call login() first.");
    }

    const response: any = await this.request("/auth/api-key", {
      method: "POST",
      body: { label }
    });

    if (response.api_key) {
      this.apiKey = response.api_key;
    }

    return response;
  }

  /**
   * Manually set API key (e.g., from environment variable)
   */
  setApiKey(apiKey: string): void {
    this.apiKey = apiKey;
  }

  private async request<T>(
    endpoint: string,
    options: RequestOptions = {}
  ): Promise<T> {
    const url = `${this.baseUrl}${endpoint}`;
    const controller = new AbortController();
    const timeoutId = setTimeout(
      () => controller.abort(),
      options.timeout || this.timeout
    );

    const fetchOptions: RequestInit = {
      method: options.method || "GET",
      headers: options.headers || this.getHeaders(),
      signal: controller.signal
    };

    if (options.body) {
      fetchOptions.body = JSON.stringify(options.body);
    }

    try {
      const response = await fetch(url, fetchOptions);

      if (!response.ok) {
        const error = await response.text();
        throw new Error(`HTTP ${response.status}: ${error}`);
      }

      const contentType = response.headers.get("content-type");
      if (contentType?.includes("application/json")) {
        return response.json() as Promise<T>;
      }

      return response.text() as any;
    } finally {
      clearTimeout(timeoutId);
    }
  }

  setToken(token: string): void {
    this.token = token;
  }

  // ==================== Health & Status ====================

  async health(): Promise<any> {
    return this.request("/health");
  }

  async status(): Promise<any> {
    return this.request("/status");
  }

  async systemStatus(): Promise<any> {
    return this.request("/api/system-status");
  }

  async dbPing(): Promise<any> {
    return this.request("/db/ping");
  }

  async redisPing(): Promise<any> {
    return this.request("/redis/ping");
  }

  // ==================== Ask & Neural Symphony ====================

  async ask(
    question: string,
    context?: string,
    includeDetails: boolean = true
  ): Promise<any> {
    return this.request("/api/ask", {
      method: "POST",
      headers: this.getHeaders(),
      body: {
        question,
        context,
        include_details: includeDetails
      }
    });
  }

  async neuralSymphony(saveTo?: string): Promise<ArrayBuffer> {
    const headers = this.getHeaders({ Accept: "audio/wav" });
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);

    try {
      const response = await fetch(`${this.baseUrl}/neural-symphony`, {
        headers,
        signal: controller.signal
      });

      if (!response.ok) {
        throw new Error(`HTTP ${response.status}`);
      }

      const buffer = await response.arrayBuffer();

      if (saveTo && typeof window === "undefined") {
        // Node.js environment
        const fs = await import("fs");
        fs.writeFileSync(saveTo, Buffer.from(buffer));
      }

      return buffer;
    } finally {
      clearTimeout(timeoutId);
    }
  }

  // ==================== Uploads & Processing ====================

  async uploadEeg(filePath: string): Promise<any> {
    if (typeof window !== "undefined") {
      throw new Error("Use uploadEegBrowser() in browser environment");
    }

    const fs = await import("fs");
    const FormData = (await import("form-data")).default;
    const form = new FormData();
    form.append("file", fs.createReadStream(filePath));

    return this.request("/api/uploads/eeg/process", {
      method: "POST",
      headers: {
        ...this.getHeaders(),
        ...form.getHeaders()
      },
      body: form
    });
  }

  async uploadEegBrowser(file: File): Promise<any> {
    const formData = new FormData();
    formData.append("file", file);

    const response = await fetch(`${this.baseUrl}/api/uploads/eeg/process`, {
      method: "POST",
      headers: this.getHeaders(),
      body: formData
    });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}`);
    }

    return response.json();
  }

  async uploadAudio(filePath: string): Promise<any> {
    if (typeof window !== "undefined") {
      throw new Error("Use uploadAudioBrowser() in browser environment");
    }

    const fs = await import("fs");
    const FormData = (await import("form-data")).default;
    const form = new FormData();
    form.append("file", fs.createReadStream(filePath));

    return this.request("/api/uploads/audio/process", {
      method: "POST",
      headers: {
        ...this.getHeaders(),
        ...form.getHeaders()
      },
      body: form
    });
  }

  async uploadAudioBrowser(file: File): Promise<any> {
    const formData = new FormData();
    formData.append("file", file);

    const response = await fetch(`${this.baseUrl}/api/uploads/audio/process`, {
      method: "POST",
      headers: this.getHeaders(),
      body: formData
    });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}`);
    }

    return response.json();
  }

  // ==================== Brain Engine ====================

  async brainYoutubeInsight(videoId: string): Promise<any> {
    return this.request(`/brain/youtube/insight?video_id=${videoId}`);
  }

  async brainEnergyCheck(filePath: string): Promise<any> {
    if (typeof window !== "undefined") {
      throw new Error("Use brainEnergyCheckBrowser() in browser");
    }

    const fs = await import("fs");
    const FormData = (await import("form-data")).default;
    const form = new FormData();
    form.append("file", fs.createReadStream(filePath));

    return this.request("/brain/energy/check", {
      method: "POST",
      headers: {
        ...this.getHeaders(),
        ...form.getHeaders()
      },
      body: form
    });
  }

  async brainEnergyCheckBrowser(file: File): Promise<any> {
    const formData = new FormData();
    formData.append("file", file);

    const response = await fetch(`${this.baseUrl}/brain/energy/check`, {
      method: "POST",
      headers: this.getHeaders(),
      body: formData
    });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}`);
    }

    return response.json();
  }

  async brainHarmony(filePath: string): Promise<any> {
    if (typeof window !== "undefined") {
      throw new Error("Use brainHarmonyBrowser() in browser");
    }

    const fs = await import("fs");
    const FormData = (await import("form-data")).default;
    const form = new FormData();
    form.append("file", fs.createReadStream(filePath));

    return this.request("/brain/harmony", {
      method: "POST",
      headers: {
        ...this.getHeaders(),
        ...form.getHeaders()
      },
      body: form
    });
  }

  async brainHarmonyBrowser(file: File): Promise<any> {
    const formData = new FormData();
    formData.append("file", file);

    const response = await fetch(`${this.baseUrl}/brain/harmony`, {
      method: "POST",
      headers: this.getHeaders(),
      body: formData
    });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}`);
    }

    return response.json();
  }

  async brainCortexMap(): Promise<any> {
    return this.request("/brain/cortex-map");
  }

  async brainTemperature(): Promise<any> {
    return this.request("/brain/temperature");
  }

  async brainQueue(): Promise<any> {
    return this.request("/brain/queue");
  }

  async brainThreads(): Promise<any> {
    return this.request("/brain/threads");
  }

  async brainNeuralLoad(): Promise<any> {
    return this.request("/brain/neural-load");
  }

  async brainErrors(): Promise<any[]> {
    return this.request("/brain/errors");
  }

  async brainRestart(): Promise<any> {
    return this.request("/brain/restart", {
      method: "POST",
      headers: this.getHeaders()
    });
  }

  // ==================== ALBA Data Collection ====================

  async albaStatus(): Promise<any> {
    return this.request("/api/alba/status");
  }

  async albaStreamsStart(
    streamId: string,
    samplingRateHz: number = 256,
    channels?: string[],
    description?: string,
    metadata?: Record<string, any>
  ): Promise<any> {
    return this.request("/api/alba/streams/start", {
      method: "POST",
      headers: this.getHeaders(),
      body: {
        stream_id: streamId,
        sampling_rate_hz: samplingRateHz,
        channels: channels || ["C3", "C4"],
        description,
        metadata: metadata || {}
      }
    });
  }

  async albaStreamsStop(streamId: string): Promise<any> {
    return this.request(`/api/alba/streams/${streamId}/stop`, {
      method: "POST",
      headers: this.getHeaders()
    });
  }

  async albaStreamsList(): Promise<any> {
    return this.request("/api/alba/streams");
  }

  async albaStreamsData(streamId: string, limit: number = 100): Promise<any> {
    return this.request(
      `/api/alba/streams/${streamId}/data?limit=${limit}`
    );
  }

  async albaMetrics(): Promise<any> {
    return this.request("/api/alba/metrics");
  }

  async albaHealth(): Promise<any> {
    return this.request("/api/alba/health");
  }
}

// ==================== Example Usage ====================

export const exampleUsage = async () => {
  // Initialize client
  const client = new ClisonixClient({
    baseUrl: "https://api.clisonix.com",
    token: "your-jwt-token-here"
  });

  try {
    // Check health
    const health = await client.health();
    console.log("✓ Health:", health.status);

    // Ask a question
    const answer = await client.ask("Çfarë është Clisonix?");
    console.log("✓ Answer:", answer.answer.substring(0, 100) + "...");

    // Start data stream
    const stream = await client.albaStreamsStart(
      "demo-stream",
      256,
      ["C3", "C4", "Pz"]
    );
    console.log("✓ Stream started:", stream.stream_id);

    // Get stream data
    const data = await client.albaStreamsData(stream.stream_id, 10);
    console.log(`✓ Retrieved ${data.data_points.length} data points`);

    // Stop stream
    const stopped = await client.albaStreamsStop(stream.stream_id);
    console.log("✓ Stream stopped:", stopped.status);
  } catch (error) {
    console.error("✗ Error:", error);
  }
};

export default ClisonixClient;

File: clisonix_sdk.py
"""
Clisonix Cloud SDK – Python Client
Part of UltraWebThinking / Euroweb
"""

import asyncio
from typing import Any, Dict, Optional
import requests
from pathlib import Path


class ClisonixClient:
    """Synchronous Clisonix API client"""

    def __init__(
        self,
        base_url: str = "https://api.clisonix.com",
        token: Optional[str] = None,
        timeout: int = 30
    ):
        self.base_url = base_url.rstrip("/")
        self.token = token
        self.refresh_token: Optional[str] = None
        self.api_key: Optional[str] = None
        self.timeout = timeout
        self.session = requests.Session()

    def _headers(self, json_content: bool = True) -> Dict[str, str]:
        headers = {
            "User-Agent": "Clisonix-Python-SDK/1.0"
        }
        if json_content:
            headers["Accept"] = "application/json"
            headers["Content-Type"] = "application/json"
        if self.token:
            headers["Authorization"] = f"Bearer {self.token}"
        if self.api_key:
            headers["X-API-Key"] = self.api_key
        return headers

    def set_token(self, token: str) -> None:
        """Set or update the authentication token"""
        self.token = token

    def set_api_key(self, api_key: str) -> None:
        """Set or update the API key"""
        self.api_key = api_key

    # ==================== Authentication ====================

    def login(self, email: str, password: str) -> Dict[str, Any]:
        """POST /auth/login – Authenticate with email & password"""
        payload = {"email": email, "password": password}
        r = self.session.post(
            f"{self.base_url}/auth/login",
            json=payload,
            headers=self._headers(),
            timeout=self.timeout
        )
        r.raise_for_status()
        data = r.json()
        
        # Store tokens
        self.token = data.get("token")
        self.refresh_token = data.get("refresh_token")
        if api_key := data.get("api_key"):
            self.api_key = api_key
        
        return data

    def refresh(self) -> Dict[str, Any]:
        """POST /auth/refresh – Refresh the JWT token"""
        if not self.refresh_token:
            raise RuntimeError("No refresh_token stored. Call login() first.")
        
        payload = {"refresh_token": self.refresh_token}
        r = self.session.post(
            f"{self.base_url}/auth/refresh",
            json=payload,
            headers=self._headers(),
            timeout=self.timeout
        )
        r.raise_for_status()
        data = r.json()
        
        # Update token
        self.token = data.get("token")
        return data

    def create_api_key(self, label: str) -> Dict[str, Any]:
        """POST /auth/api-key – Create a new API key"""
        if not self.token:
            raise RuntimeError("Authentication required. Call login() first.")
        
        payload = {"label": label}
        r = self.session.post(
            f"{self.base_url}/auth/api-key",
            json=payload,
            headers=self._headers(),
            timeout=self.timeout
        )
        r.raise_for_status()
        data = r.json()
        
        # Store the API key
        if api_key := data.get("api_key"):
            self.api_key = api_key
        
        return data

    # ==================== Health & Status ====================

    def health(self) -> Dict[str, Any]:
        """GET /health – System health check"""
        r = self.session.get(
            f"{self.base_url}/health",
            headers=self._headers(),
            timeout=self.timeout
        )
        r.raise_for_status()
        return r.json()

    def status(self) -> Dict[str, Any]:
        """GET /status – Full system status"""
        r = self.session.get(
            f"{self.base_url}/status",
            headers=self._headers(),
            timeout=self.timeout
        )
        r.raise_for_status()
        return r.json()

    def system_status(self) -> Dict[str, Any]:
        """GET /api/system-status – System status proxy"""
        r = self.session.get(
            f"{self.base_url}/api/system-status",
            headers=self._headers(),
            timeout=self.timeout
        )
        r.raise_for_status()
        return r.json()

    def db_ping(self) -> Dict[str, Any]:
        """GET /db/ping – Database connectivity test"""
        r = self.session.get(
            f"{self.base_url}/db/ping",
            headers=self._headers(),
            timeout=self.timeout
        )
        r.raise_for_status()
        return r.json()

    def redis_ping(self) -> Dict[str, Any]:
        """GET /redis/ping – Redis connectivity test"""
        r = self.session.get(
            f"{self.base_url}/redis/ping",
            headers=self._headers(),
            timeout=self.timeout
        )
        r.raise_for_status()
        return r.json()

    # ==================== Ask & Neural Symphony ====================

    def ask(
        self,
        question: str,
        context: Optional[str] = None,
        include_details: bool = True
    ) -> Dict[str, Any]:
        """POST /api/ask – Query the AI assistant"""
        payload = {
            "question": question,
            "context": context,
            "include_details": include_details
        }
        r = self.session.post(
            f"{self.base_url}/api/ask",
            json=payload,
            headers=self._headers(),
            timeout=self.timeout
        )
        r.raise_for_status()
        return r.json()

    def neural_symphony(self, save_to: Optional[str] = None) -> bytes:
        """GET /neural-symphony – Generate EEG alpha wave audio"""
        headers = self._headers(json_content=False)
        headers["Accept"] = "audio/wav"
        r = self.session.get(
            f"{self.base_url}/neural-symphony",
            headers=headers,
            timeout=self.timeout
        )
        r.raise_for_status()
        
        if save_to:
            Path(save_to).write_bytes(r.content)
        
        return r.content

    # ==================== Uploads & Processing ====================

    def upload_eeg(self, file_path: str) -> Dict[str, Any]:
        """POST /api/uploads/eeg/process – Process EEG file"""
        with open(file_path, "rb") as f:
            files = {"file": f}
            r = self.session.post(
                f"{self.base_url}/api/uploads/eeg/process",
                files=files,
                headers={k: v for k, v in self._headers(json_content=False).items() if k != "Content-Type"},
                timeout=self.timeout
            )
        r.raise_for_status()
        return r.json()

    def upload_audio(self, file_path: str) -> Dict[str, Any]:
        """POST /api/uploads/audio/process – Process audio file"""
        with open(file_path, "rb") as f:
            files = {"file": f}
            r = self.session.post(
                f"{self.base_url}/api/uploads/audio/process",
                files=files,
                headers={k: v for k, v in self._headers(json_content=False).items() if k != "Content-Type"},
                timeout=self.timeout
            )
        r.raise_for_status()
        return r.json()

    # ==================== Brain Engine ====================

    def brain_youtube_insight(self, video_id: str) -> Dict[str, Any]:
        """GET /brain/youtube/insight – YouTube video analysis"""
        r = self.session.get(
            f"{self.base_url}/brain/youtube/insight",
            params={"video_id": video_id},
            headers=self._headers(),
            timeout=self.timeout
        )
        r.raise_for_status()
        return r.json()

    def brain_energy_check(self, file_path: str) -> Dict[str, Any]:
        """POST /brain/energy/check – Daily energy analysis"""
        with open(file_path, "rb") as f:
            files = {"file": f}
            r = self.session.post(
                f"{self.base_url}/brain/energy/check",
                files=files,
                headers={k: v for k, v in self._headers(json_content=False).items() if k != "Content-Type"},
                timeout=self.timeout
            )
        r.raise_for_status()
        return r.json()

    def brain_harmony(self, file_path: str) -> Dict[str, Any]:
        """POST /brain/harmony – Harmonic structure analysis"""
        with open(file_path, "rb") as f:
            files = {"file": f}
            r = self.session.post(
                f"{self.base_url}/brain/harmony",
                files=files,
                headers={k: v for k, v in self._headers(json_content=False).items() if k != "Content-Type"},
                timeout=self.timeout
            )
        r.raise_for_status()
        return r.json()

    def brain_cortex_map(self) -> Dict[str, Any]:
        """GET /brain/cortex-map – Neural architecture map"""
        r = self.session.get(
            f"{self.base_url}/brain/cortex-map",
            headers=self._headers(),
            timeout=self.timeout
        )
        r.raise_for_status()
        return r.json()

    def brain_temperature(self) -> Dict[str, Any]:
        """GET /brain/temperature – Module thermal stress"""
        r = self.session.get(
            f"{self.base_url}/brain/temperature",
            headers=self._headers(),
            timeout=self.timeout
        )
        r.raise_for_status()
        return r.json()

    def brain_restart(self) -> Dict[str, Any]:
        """POST /brain/restart – Safely restart cognitive modules"""
        r = self.session.post(
            f"{self.base_url}/brain/restart",
            headers=self._headers(),
            timeout=self.timeout
        )
        r.raise_for_status()
        return r.json()

    # ==================== ALBA Data Collection ====================

    def alba_status(self) -> Dict[str, Any]:
        """GET /api/alba/status – ALBA module status"""
        r = self.session.get(
            f"{self.base_url}/api/alba/status",
            headers=self._headers(),
            timeout=self.timeout
        )
        r.raise_for_status()
        return r.json()

    def alba_streams_start(
        self,
        stream_id: str,
        sampling_rate_hz: int = 256,
        channels: Optional[list] = None,
        description: Optional[str] = None,
        metadata: Optional[dict] = None
    ) -> Dict[str, Any]:
        """POST /api/alba/streams/start – Start data stream"""
        payload = {
            "stream_id": stream_id,
            "sampling_rate_hz": sampling_rate_hz,
            "channels": channels or ["C3", "C4"],
            "description": description,
            "metadata": metadata or {}
        }
        r = self.session.post(
            f"{self.base_url}/api/alba/streams/start",
            json=payload,
            headers=self._headers(),
            timeout=self.timeout
        )
        r.raise_for_status()
        return r.json()

    def alba_streams_stop(self, stream_id: str) -> Dict[str, Any]:
        """POST /api/alba/streams/{stream_id}/stop – Stop data stream"""
        r = self.session.post(
            f"{self.base_url}/api/alba/streams/{stream_id}/stop",
            headers=self._headers(),
            timeout=self.timeout
        )
        r.raise_for_status()
        return r.json()

    def alba_streams_list(self) -> Dict[str, Any]:
        """GET /api/alba/streams – List active streams"""
        r = self.session.get(
            f"{self.base_url}/api/alba/streams",
            headers=self._headers(),
            timeout=self.timeout
        )
        r.raise_for_status()
        return r.json()

    def alba_streams_data(self, stream_id: str, limit: int = 100) -> Dict[str, Any]:
        """GET /api/alba/streams/{stream_id}/data – Get stream data"""
        r = self.session.get(
            f"{self.base_url}/api/alba/streams/{stream_id}/data",
            params={"limit": limit},
            headers=self._headers(),
            timeout=self.timeout
        )
        r.raise_for_status()
        return r.json()

    def alba_metrics(self) -> Dict[str, Any]:
        """GET /api/alba/metrics – Collection metrics"""
        r = self.session.get(
            f"{self.base_url}/api/alba/metrics",
            headers=self._headers(),
            timeout=self.timeout
        )
        r.raise_for_status()
        return r.json()

    def alba_health(self) -> Dict[str, Any]:
        """GET /api/alba/health – ALBA health check"""
        r = self.session.get(
            f"{self.base_url}/api/alba/health",
            headers=self._headers(),
            timeout=self.timeout
        )
        r.raise_for_status()
        return r.json()

    def close(self) -> None:
        """Close the session"""
        self.session.close()

    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.close()


# ==================== Example Usage ====================

if __name__ == "__main__":
    client = ClisonixClient(base_url="https://api.clisonix.com")

    try:
        # 1. Login
        login_data = client.login("user@example.com", "password")
        print(f"✓ Logged in. Token: {login_data['token'][:20]}...")
        print(f"✓ Refresh token: {login_data['refresh_token'][:20]}...")

        # 2. Check health
        health = client.health()
        print(f"✓ Health: {health['status']}")

        # 3. Ask a question
        answer = client.ask("Çfarë është Clisonix?", "neuro-audio platform")
        print(f"✓ Answer: {answer['answer'][:100]}...")

        # 4. Refresh token (optional)
        refreshed = client.refresh()
        print(f"✓ Token refreshed: {refreshed['token'][:20]}...")

        # 5. Create API key
        api_key_response = client.create_api_key("my-server-integration")
        print(f"✓ API Key created: {api_key_response['api_key'][:20]}...")

    except Exception as e:
        print(f"✗ Error: {e}")
    finally:
        client.close()
